packages/dev/core/src/Materials/shaderMaterial.ts

```diff
@@ -35,6 +35,45 @@ import {
 
 const onCreatedEffectParameters = { effect: null as unknown as Effect, subMesh: null as unknown as Nullable<SubMesh> };
 
+/**
+ * Defines the route to the shader code. The priority is as follows:
+ *  * object: `{ vertexSource: "vertex shader code string", fragmentSource: "fragment shader code string" }` for directly passing the shader code
+ *  * object: `{ vertexElement: "vertexShaderCode", fragmentElement: "fragmentShaderCode" }`, used with shader code in script tags
+ *  * object: `{ vertex: "custom", fragment: "custom" }`, used with `Effect.ShadersStore["customVertexShader"]` and `Effect.ShadersStore["customFragmentShader"]`
+ *  * string: `"./COMMON_NAME"`, used with external files COMMON_NAME.vertex.fx and COMMON_NAME.fragment.fx in index.html folder.
+ */
+export type IShaderPath =
+    | {
+          /**
+           * Directly pass the shader code
+           */
+          vertexSource?: string;
+          /**
+           * Directly pass the shader code
+           */
+          fragmentSource?: string;
+
+          /**
+           * Used with Effect.ShadersStore. If the `vertex` is set to `"custom`, then
+           * Babylon.js will read from Effect.ShadersStore["customVertexShader"]
+           */
+          vertex?: string;
+          /**
+           * Used with Effect.ShadersStore. If the `fragment` is set to `"custom`, then
+           * Babylon.js will read from Effect.ShadersStore["customFragmentShader"]
+           */
+          fragment?: string;
+          /**
+           * Used with shader code in script tags
+           */
+          vertexElement?: string;
+          /**
+           * Used with shader code in script tags
+           */
+          fragmentElement?: string;
+      }

```

Comment:
See my comment in the thread: add `vertexToken?: string` in the object, but document is as `/* @internal */` so that it's not visible to users.